src/lem/circuit.rs

```diff
@@ -0,0 +1,927 @@
+        let preallocated_hash2_slots = LEM::allocate_slots(
+            cs,
+            &frame.preimages.hash2_ptrs,
+            SlotType::Hash2,
+            slots_count.hash2,
+            store,
+        )?;
+
+        let preallocated_hash3_slots = LEM::allocate_slots(
+            cs,
+            &frame.preimages.hash3_ptrs,
+            SlotType::Hash3,
+            slots_count.hash3,
+            store,
+        )?;
+
+        let preallocated_hash4_slots = LEM::allocate_slots(
+            cs,
+            &frame.preimages.hash4_ptrs,
+            SlotType::Hash4,
+            slots_count.hash4,
+            store,
+        )?;
+
+        struct Globals<'a, F: LurkField> {
+            store: &'a mut Store<F>,
+            frame: &'a Frame<F>,
+            global_allocator: &'a mut GlobalAllocator<F>,
+            allocated_ptrs: HashMap<AString, AllocatedPtr<F>>,
+            preallocated_outputs: Vec<AllocatedPtr<F>>,
+            preallocated_hash2_slots: Vec<(Vec<AllocatedNum<F>>, AllocatedNum<F>)>,
+            preallocated_hash3_slots: Vec<(Vec<AllocatedNum<F>>, AllocatedNum<F>)>,
+            preallocated_hash4_slots: Vec<(Vec<AllocatedNum<F>>, AllocatedNum<F>)>,
+        }
+
+        fn recurse<F: LurkField, CS: ConstraintSystem<F>>(
+            cs: &mut CS,
+            block: &LEMCTL,
+            concrete_path: Boolean,
+            slots_count: &mut SlotsCounter,
+            g: &mut Globals<'_, F>,
+        ) -> Result<()> {
+            match block {
+                LEMCTL::Return(output_vars) => {
+                    for (i, output_var) in output_vars.iter().enumerate() {
+                        let Some(allocated_ptr) = g.allocated_ptrs.get(output_var.name()) else {
+                            bail!("{output_var} not allocated")
+                        };
+
+                        allocated_ptr
+                            .implies_ptr_equal(
+                                &mut cs.namespace(|| {
+                                    format!("implies_ptr_equal {output_var} (output_var {i})")
+                                }),
+                                &concrete_path,
+                                &g.preallocated_outputs[i],
+                            )
+                            .with_context(|| "couldn't constrain `implies_ptr_equal`")?;
+                    }
+                    Ok(())
+                }
+                LEMCTL::MatchTag(match_ptr, cases) => {
+                    let allocated_match_tag = match g.allocated_ptrs.get(match_ptr.name()) {
+                        Some(allocated_match_ptr) => allocated_match_ptr.tag().clone(),
+                        None => bail!("{match_ptr} not allocated"),
+                    };
+                    let mut concrete_path_vec = Vec::new();
+                    for (tag, op) in cases {
+                        let allocated_has_match = alloc_equal_const(
+                            &mut cs.namespace(|| format!("{tag}.alloc_equal_const")),
```

Comment:
`cases` must not contain duplicate tags or else this will result in a synthesis error (duplicated path).